src/py21cmfast/cli.py

```diff
@@ -1,827 +1,733 @@
-    ics = InitialConditions(inputs=inputs)
+    outcfg: Annotated[
+        Path,
+        Parameter(validator=(vld.Path(file_okay=False, dir_okay=False, ext=("toml",)))),
+    ] = None
+    "A filepath where the full configuration TOML of the run can be written. Random hash if not specified."
+
+    verbosity: Annotated[
+        Literal["CRITICAL", "ERROR", "WARNING", "INFO", "DEBUG"],
+        Parameter(alias=("-v", "--v")),
+    ] = "WARNING"
+
+    progress: bool = True
+
+
+def _param_cls_factory(cls):
+    out = attrs.make_class(
+        cls.__name__,
+        {
+            fld.name: attrs.field(default=None, type=fld.type)
+            for fld in attrs.fields(cls)
+        },
+        kw_only=True,
+        frozen=True,
+    )
+    out.__doc__ = cls.__doc__.replace(":class:", "")
+    return out
+
+
+_AstroOptions = _param_cls_factory(AstroOptions)
+_AstroParams = _param_cls_factory(AstroParams)
+_CosmoParams = _param_cls_factory(CosmoParams)
+_MatterOptions = _param_cls_factory(MatterOptions)
+_SimulationOptions = _param_cls_factory(SimulationOptions)
+
+
+@Parameter(name="*")
+@dataclass
+class Parameters:
+    """A trimmed-down version of InputParameters with all defaults of None."""
+
+    simulation_options: Annotated[
+        _SimulationOptions, Parameter(name="*", group="SimulationOptions")
+    ] = _SimulationOptions()
+    astro_options: Annotated[
+        _AstroOptions, Parameter(name="*", group="AstroOptions")
+    ] = _AstroOptions()
+    astro_params: Annotated[_AstroParams, Parameter(name="*", group="AstroParams")] = (
+        _AstroParams()
+    )
+    cosmo_params: Annotated[_CosmoParams, Parameter(name="*", group="CosmoParams")] = (
+        _CosmoParams()
+    )
+    matter_options: Annotated[
+        _MatterOptions, Parameter(name="*", group="MatterOptions")
+    ] = _MatterOptions()
 
-    perturb_field(
-        redshift=redshift,
-        inputs=inputs,
-        initial_conditions=ics.read(direc),
-        regenerate=regen,
-        write=True,
-        direc=direc,
+
+def _get_inputs(
+    options: RunParams | ParameterSelection, params: Parameters
+) -> tuple[InputParameters, bool]:
+    pselect = options.param_selection if isinstance(options, RunParams) else options
+
+    seed = getattr(options, "seed", 42)
+
+    if pselect.param_file is not None:
+        inputs = InputParameters.from_template(pselect.param_file, random_seed=seed)
+    elif pselect.template == "defaults":
+        inputs = InputParameters(random_seed=seed)
+    else:
+        inputs = InputParameters.from_template(pselect.template, random_seed=seed)
+
+    # kwargs from params:
+    kwargs = {}
+    for field in fields(Parameters):
+        this = getattr(params, field.name)
+        kwargs |= {
+            name: val for name, val in attrs.asdict(this).items() if val is not None
+        }
+
+    new = inputs.evolve_input_structs(**kwargs)
+    modified = new != inputs
+    return new, modified
+
```

Comment:
Just for my understanding, the `_param_cls_factory` returns a new class based on the parameter class that holds all the fields but where they are all optional and default `None`. These are grouped into the `Parameters` class and combined with the other options in `RunOptions`, which contains a template. `Parameters` overwrites all parameters from the template given in `RunOptions`

Reply:
Yep! Basically, I wanted to make it possible to set any parameter in any of the `InputStruct` classes (`AstroParams` etc), but I didn't want to re-write all the parameters that exist in all of the `InputStruct` subclasses (`AstroParams` etc). At the same time, I wanted the behaviour to be such that when the user explicitly sets such a parameter, it overwrites whatever is given in `--template` or `--param-file`.

So, I couldn't just use the `InputStruct` classes directly, because they all have defaults, and so I couldn't tell if they were "set" or not, so they would always overwrite the template. Thus, I create new dummy classes that just have all the parameters as well as the docstrings, but with default of None.